b288702/download_initial_generation_prereqs.sh

```diff
@@ -0,0 +1,95 @@
+#!/usr/bin/env bash
+
+# This script downloads the files needed to verify a the first contribution.
+#
+# Inputs are a proof type and a sector size.
+set -e
+
+script_name=$(basename "$0")
+
+if [ "${#}" -ne 3 ]; then
+    echo "Download files needed to verify the first contribution."
+    echo ""
+    echo "Usage: ${script_name} {sdr|window|winning} {32|64}"
+    exit 1
+fi
+
+if ! command -v b2sum >/dev/null 2>&1
+then
+    echo "ERROR: 'b2sum' needs to be installed."
+    exit 1
+fi
+
+if ! command -v curl >/dev/null 2>&1
+then
+    echo "ERROR: 'curl' needs to be installed."
+    exit 1
+fi
+
+proof="$1"
+sector_size="$2"
+
+magenta='\u001b[35;1m'
+red='\u001b[31;1m'
+off='\u001b[0m'
+
+function log() {
+    echo -e "${magenta}[${script_name}]${off} $1"
+}
+
+function error() {
+    echo -e "${magenta}[${script_name}] ${red}error:${off} $1"
+}
+
+if [[ $proof != 'winning' && $proof != 'sdr' && $proof != 'window' ]]; then
+    error "invalid proof: '${proof}'"
+    exit 1
+fi
+
+if [[ $sector_size != '32' && $sector_size != '64' ]]; then
+    error "invalid sector-size: '${sector_size}'"
+    exit 1
+fi
+
+# Once everything is pinned to IPFS, it will be a single location for all
+# the files
+url_base='https://trusted-setup.s3.amazonaws.com/phase2-mainnet'
+url_base_phase1='https://trusted-setup.s3.amazonaws.com/challenge19'
+
+# Get the file containing checksums of the parameter files.
+if [[ ! -f './b288702.b2sums' ]]; then
+    log "downloading checksums for parameters: b288702.b2sums"
+    curl --progress-bar -O "${url_base}/b288702.b2sums"
+fi
+log 'verifying parameters checksums file checksum'
+echo "7931ca92df34bf0b6217692daaf2d92135fceb6caae344b10712ee997717cc612435b0a6e1e61325d5abaa62044b6f6359fd44bbe3dc4e111536bcad43c2e0ec  b288702.b2sums" | b2sum -c
+
+if [[ $proof == 'winning' ]]; then
+    phase1_file='phase1radix2m19'
+    phase1_checksum='4a3b6930739967248fee48dbf43e27ee907ab3780132e21d4c7fe37fcebdc87352f1495795178c27799828db9da3696eb6ef19054404b23ec4994883877d96f8'
+else
+    phase1_file='phase1radix2m27'
+    phase1_checksum='8a5d4e211e3a9817dcdc7d345a25338f261d0b52ab188661dfcb6bada9f2f5ac76925521621d8f87b5b680105973f4b48fb1ec68f65ebf8fdbccbd8d4891e6e9'
+fi
+
+if [[ ! -f ${phase1_file} ]]; then
+    log "downloading Phase 1 parameters: ${phase1_file}"
+    curl --progress-bar -O "${url_base_phase1}/${phase1_file}"
+fi
+log 'verifying Phase 1 checksum'
+echo "${phase1_checksum} ${phase1_file}" | b2sum -c
+
+# Download the parameters of the first contribution
+file="${proof}_poseidon_${sector_size}gib_b288702_1_small_raw"
+if [[ ! -f ${file} ]]; then
+    log "downloading params: ${file}"
+    curl --progress-bar -O "${url_base}/${file}"
+fi
```

Comment:
I guess for greatest consistency, this would still be in `download_prereqs_contrib.sh`. I think I can live with it here, though. @ribasushi ?

Reply:
@porcuquine actually I requested it to be separate. Needing to download it on *every* step is wasteful ( the matrix is *huge* ). Since it is needed on step 1 only - I asked @vmx to download it only when I will be generating "0_large"

Reply:
@porcuquine Do you want to have a `./download_prereqs_contrib.sh` call also in the inital step? Like this:

./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && **./download_prereqs_contrib.sh winning 32 1** && ./verify_contrib.sh winning 32 1

Instead of what we are currently doing:

./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && ./verify_contrib.sh winning 32 1

Reply:
@vmx the more distinct steps there are in the ` ... && ...` list - the better
There already is one "magic script" of ./verify-all. Further obscuring at intermediate steps is not reasonable.

Reply:
> @vmx the more distinct steps there are in the ` ... && ...` list - the better

Which means? Adding the additional `./download_prereqs_contrib.sh winning 32 1` would mean that this script would get a special case.

Reply:
I think this is ok / desired.
@porcuquine thoughts?

Reply:
@ribasushi 
> Needing to download it on every step is wasteful ( the matrix is huge ). Since it is needed on step 1 only - I asked @vmx to download it only when I will be generating "0_large"

I am not sure I follow. I think the confusion might be that the lines I intended to comment on are the ones which download the step 1 file:
`file="${proof}_poseidon_${sector_size}gib_b288702_1_small_raw"`.

NOTE: I am catching up to this conversation and will try to answer things in order, so my first responses may be outdated.

Reply:
> Do you want to have a ./download_prereqs_contrib.sh call also in the inital step? Like this:
>
> ```./download_initial_generation_prereqs.sh winning 32 && ./generate_initial.sh winning 32 && ./download_prereqs_contrib.sh winning 32 1 && ./verify_contrib.sh winning 32 1```

Yes, that is what I was suggesting. I think this is the organization which was logically implied by all the discussions we had yesterday. We just didn't quite fully break it down on the first pass, and I didn't notice until I made that review comment.

Reply:
> Which means? Adding the additional ./download_prereqs_contrib.sh winning 32 1 would mean that this script would get a special case.

> I think this is ok / desired.

Yes, agreed. For clarity, the special case is that we should never download the 'before' (0) params when verifying the first (1) contribution. We should always use the generated ones. We can still verify the checksum as usual and fail if the generated params were not as expected.

Reply:
> We should always use the generated ones.

Not sure if this is the right time/place: but perhaps ./phase2 shouldn't be so "magical/helpful" and instead require all its arguments to be explicitly specified.

Not a request, just thinking out-loud

Reply:
I think I got the idea, I'll do a PR.